--- /Users/brantley/nixpkgs/./pkgs/build-support/buildenv/default.nix	2024-09-02 07:29:20
+++ pkgs/flox-buildenv/buildenv.nix	2024-09-13 14:10:21
@@ -8,11 +8,15 @@
     src = ./builder.pl;
     inherit (builtins) storeDir;
   };
+  build_closures_jq = ./build-closures.jq;
 in
 
 lib.makeOverridable
 ({ name
 
+, # The path to the flox "activation-scripts" package.
+  activationScripts
+
 , # The manifest file (if any).  A symlink $out/manifest will be
   # created to it.
   manifest ? ""
@@ -50,11 +54,24 @@
 , meta ? {}
 }:
 
+let
+    # Unlike the nixpkgs buildEnv, the Flox one has multiple outputs,
+    # the usual builtins.buildenv() output, a "develop" output, and
+    # a runtime closure for each of the manifest builds.
+    manifestBuilds = if manifest == "" then [] else (
+      let _manifest = builtins.fromJSON (builtins.readFile manifest);
+      in builtins.attrNames _manifest.manifest.build
+    );
+    runtimePkgPrefix = "build_";
+    manifestBuildRuntimeOutputs = map (s: runtimePkgPrefix + s) manifestBuilds;
+    outputs = ["out" "develop"] ++ manifestBuildRuntimeOutputs;
+
+in
 runCommand name
   rec {
     inherit manifest ignoreCollisions checkCollisionContents passthru
             meta pathsToLink extraPrefix postBuild
-            nativeBuildInputs buildInputs;
+            nativeBuildInputs buildInputs outputs;
     pkgs = builtins.toJSON (map (drv: {
       paths =
         # First add the usual output(s): respect if user has chosen explicitly,
@@ -72,10 +89,54 @@
     }) paths);
     preferLocalBuild = true;
     allowSubstitutes = false;
-    # XXX: The size is somewhat arbitrary
-    passAsFile = if builtins.stringLength pkgs >= 128*1024 then [ "pkgs" ] else [ ];
+    # Nix "*Path" environment variables are automatically created by
+    # way of the derivation `passAsFile` attribute as described in:
+    #
+    # https://nix.dev/manual/nix/2.18/language/advanced-attributes#adv-attr-passAsFile
+    #
+    # The following causes `pkgsPath` to always be set.
+    passAsFile = [ "pkgs" ];
   }
   ''
     ${buildPackages.perl}/bin/perl -w ${builder}
+
+    # The `builder.pl` script expects to receive the list of packages by
+    # way of one of the `pkgsPath` or `pkgs` environment variables. Explicitly
+    # set `pkgsPath` to pass our modified when building flox environments.
+
+    # Assert $pkgsPath is set.
+    if [ -z "$pkgsPath" ]; then
+      echo "Error: \$pkgsPath is not set." >&2
+      exit 1
+    fi
+
+    # Add the activation scripts package to the list of packages and
+    # build the "develop" output.
+    tmppkgs=$(mktemp)
+    ${buildPackages.jq}/bin/jq -c -r '
+      . + [{
+        "paths": [ "${activationScripts}" ],
+        "priority": 1
+      }]
+    ' $pkgsPath > $tmppkgs
+    out=$develop pkgsPath=$tmppkgs FLOX_RECURSIVE_LINK=1 \
+      ${buildPackages.perl}/bin/perl -w ${builder}
+
+    # Iterate over manifest builds creating closures for each build as
+    # specified in the manifest.
+    for buildRuntimeOutput in ${builtins.toString manifestBuildRuntimeOutputs}; do
+      build="''${buildRuntimeOutput#${runtimePkgPrefix}}"
+      ${buildPackages.jq}/bin/jq -c -r -f ${build_closures_jq} \
+        --arg activationScripts ${activationScripts} \
+        --arg build $build \
+        --arg system ${builtins.currentSystem} \
+        ${manifest} > $tmppkgs
+      out="''${!buildRuntimeOutput}" pkgsPath=$tmppkgs FLOX_RECURSIVE_LINK=1 \
+        ${buildPackages.perl}/bin/perl -w ${builder}
+    done
+
+    # Clean up.
+    rm $tmppkgs
+
     eval "$postBuild"
   '')
